### DISCUSSION TITLE
Framework Events

### DISCUSSION THREAD

**Discussion by mikecasas** [oqtane/oqtane.framework#886] [Ideas]
Would the framework benefit from an event system?  I'm not suggesting something too complex, but would the ability to have subscribers listening for events help anyone else?

**Comment by hishamco** [comment]
I like the idea of having simple Event Bus System that allow the users to subscribe and get notifying when the event occurs, if sound looks good to @sbwalker I can suggest my opinion about it

Thanks for bringing this

**Reply by mikecasas** [reply]
I would second an event bus, but I didn't want to introduce it as it may complicate a developer just starting out with oqtane.

**Reply by hishamco** [reply]
Simply the events should be attached with the available services such as User Created, Page Loaded ... etc

**Reply by mikesmeltzer** [reply]
This would be great, imo. Items like mentioned above plus things like user logged in, module added to a page (moved, deleted, etc.), page deleted, etc..

**Comment by sturlath** [comment]
have you seen this one? https://github.com/cpear/BlazorComponentBus

**Reply by mikecasas** [reply]
Thanks! I just did a quick read, but it seems that project is for communication within components.  I was thinking more framework wide and more backend - not necessarily between components. 

But yes some events would trigger UI components.

**Comment by sbwalker** [comment]
Yes, I think this would be a useful feature for the framework. I am generally not in favor of implementing features in any application unless there is a use case which implements it - however in the case of Oqtane there is already a fairly critical use case which needs this functionality. 

There is a concept of a SyncEvent in the framework already which is essentially a simple event bus used to communicate from the server to the client that an entity was modified so that the router knows that it needs to perform a reload ( as some entities are essentially "cached" client-side in PageState and ModuleState objects ). The current approach works however it is not extensible and it also does not have any persistence capability.

Based on the client/server architecture of Oqtane, an event system would need to be client/server as well. It would need to be able to notify clients that an event has occurred on the server. It should also be persistent so that if the server is restarted, clients can still get the notifications after the restart. It would need to support core framework use case as well as third party module use cases. It also needs to be high performance so as not to degrade the overall performance of the system. It should not require any complicated configuration for development or deployment. And preferably it should be a native solution to align with the philosophy of minimizing third party dependencies. Also the framework APIs would need to have applicable events implemented so there is complete coverage.

**Reply by mikecasas** [reply]
> And preferably it should be a native solution to align with the philosophy of minimizing third party dependencies.

So are you against using something like RabbitMQ?

**Reply by sbwalker** [reply]
We use RabbitMQ at work for many projects and it requires complicated configuration and is difficult to deploy.

**Reply by sbwalker** [reply]
Note that is it possible that the use case I identified may not be the same use case that other people are interested in having support for. The use case I explained is focused on changes on the server being propagated to client browsers for synchronization purposes. Communication between client and server is usually best served by a web socket solution. However exposing events such as User Created, Page Deleted, etc... may actually be a server-centric scenario where server components can subscribe to messages published by other server components. Server to server communication is often provided by web hooks ( a lot of popular APIs such as Github, etc... provide web hook support ). It would be useful to hear about some actual use cases where developers want to use this to determine what people are looking for.

**Reply by curelom** [reply]
I have a use case.  Some of our users will be able to send invites to people.  When those users go in and create their account, we would like to be able to look at the database and see who invited them and add certain additional roles and other database updates.

**Comment by elhajjh** [comment]
What about communication between modules on the same page.
If a module modified information that is used by the other module on the same page, we need the second module to act accordingly

**Reply by markdav-is** [reply]
I think there is cross-module messages and I'm looking for that now.  the other Option is to use something like MediatR

**Reply by sbwalker** [reply]
@markdav-is  https://www.oqtane.org/blog/!/40/inotifypropertychanged

**Reply by markdav-is** [reply]
thanks!  I found the demo video as well: https://youtu.be/Yvp_PLvJqqk?t=1321

**Comment by sbwalker** [comment]
#2449 allows developers to consume events raised by the core framework. It uses the standard the .NET EventHandler capability in conjunction with the SyncManager to raise SyncEvents (which inherit from the standard EventArgs).

The tricky part with event handling in a web context is that most services are short-lived (ie. either scoped or transient) - which is a problem because in order for you to be able to capture an event, you need to have a service which is guaranteed to be running at the precise moment when the event is fired. This is where IHostedServices can be very helpful. IHostedServices startup at the same time as your web application and run on a background thread for the lifetime of your application. This makes them a perfect candidate for capturing events. The challenge with IHostedServices are that they do not support standard dependency injection for scoped services so you need to do some extra work to instantiate those services.

The following is an example of how to capture the File Download event. You will notice that it registers the EntityChanged event handler and then inspects the SyncEvent to determine if the Entity and Action match the specific event you are interested in. You should also note that due to the multi-tenant nature of Oqtane it is important to set the tenant before calling any repository methods.


    public class FileDownloadHostedService : IHostedService
    {
        private readonly IServiceScopeFactory _serviceScopeFactory;
        private readonly ISyncManager _syncManager;

        public FileDownloadHostedService(IServiceScopeFactory serviceScopeFactory, ISyncManager syncManager)
        {
            _serviceScopeFactory = serviceScopeFactory;
            _syncManager = syncManager;
        }

        void EntityChanged(object sender, SyncEvent e)
        {
            // consume SyncEvent
            if (e.EntityName == "File" && e.Action == "Download")
            {
                using (var scope = _serviceScopeFactory.CreateScope())
                {
                    // create scoped services
                    var tenantManager = scope.ServiceProvider.GetRequiredService<ITenantManager>();
                    var fileRepository = scope.ServiceProvider.GetRequiredService<IFileRepository>();

                    // set tenant
                    tenantManager.SetTenant(e.TenantId);

                    // get file and do whatever you need
                    var file = fileRepository.GetFile(e.EntityId);
                }
            }
        }

        public Task StartAsync(CancellationToken cancellationToken)
        {
            _syncManager.EntityChanged += EntityChanged;

            return Task.CompletedTask;
        }

        public Task StopAsync(CancellationToken cancellationToken)
        {
            return Task.CompletedTask;
        }

        public void Dispose()
        {
            _syncManager.EntityChanged -= EntityChanged;
        }
    }
All of the core framework entities have been enhanced to raise Add, Update, and Delete events in this manner - which opens up many new opportunities for developers to extend the framework server-side.

Note that SyncManager already existed in prior versions of Oqtane but was only used to synchronize server-side events with client-side state. It still serves this purpose as well (with some Refresh and Reload actions that were introduced for this specific scenario).

**Comment by sbwalker** [comment]
Blog post for the new server-side event capability in 3.2.1:

https://www.oqtane.org/blog/!/53/oqtane-server-events
